{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "from messari.nfts import (\n",
    "    NFTPriceFloor,\n",
    "    NonFungible,\n",
    "    OpenSea,\n",
    ")\n",
    "\n",
    "import requests\n",
    "from string import Template\n",
    "import pandas as pd \n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Testing OpenSea API \n",
    "HEADERS = {'Accept': 'application/json'}\n",
    "ASSET_URL = Template('https://api.opensea.io/api/v1/asset/$contract/$id/')\n",
    "CONTRACT_URL = Template('https://api.opensea.io/api/v1/asset_contract/$contract')\n",
    "COLLECTION_URL = Template('https://api.opensea.io/api/v1/collection/$collection')\n",
    "STATS_URL = Template('https://api.opensea.io/api/v1/collection/$collection/stats')\n",
    "EVENTS_URL = 'https://api.opensea.io/api/v1/events'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "url = CONTRACT_URL.substitute(contract=\"0xb47e3cd837ddf8e4c57f05d70ab865de6e193bbb\")\n",
    "response = requests.get(\n",
    "    url,\n",
    "    headers=HEADERS,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<!DOCTYPE html>\\n<html lang=\"en-US\">\\n   <head>\\n      <title>Access denied</title>\\n      <meta http-equiv=\"X-UA-Compatible\" content=\"IE=Edge\" />\\n      <meta name=\"robots\" content=\"noindex, nofollow\" />\\n      <meta name=\"viewport\" content=\"width=device-width,initial-scale=1\" />\\n      <link rel=\"stylesheet\" href=\"/cdn-cgi/styles/errors.new.min.css\" media=\"screen\" />\\n      <script>\\n(function(){if(document.addEventListener&&window.XMLHttpRequest&&JSON&&JSON.stringify){var e=function(a){var c=document.getElementById(\"error-feedback-survey\"),d=document.getElementById(\"error-feedback-success\"),b=new XMLHttpRequest;a={event:\"feedback clicked\",properties:{errorCode:1020,helpful:a,version:3}};b.open(\"POST\",\"https://sparrow.cloudflare.com/api/v1/event\");b.setRequestHeader(\"Content-Type\",\"application/json\");b.setRequestHeader(\"Sparrow-Source-Key\",\"c771f0e4b54944bebf4261d44bd79a1e\");\\nb.send(JSON.stringify(a));c.classList.add(\"feedback-hidden\");d.classList.remove(\"feedback-hidden\")};document.addEventListener(\"DOMContentLoaded\",function(){var a=document.getElementById(\"error-feedback\"),c=document.getElementById(\"feedback-button-yes\"),d=document.getElementById(\"feedback-button-no\");\"classList\"in a&&(a.classList.remove(\"feedback-hidden\"),c.addEventListener(\"click\",function(){e(!0)}),d.addEventListener(\"click\",function(){e(!1)}))})}})();\\n</script>\\n\\n      <script>\\n         (function(){if(document.addEventListener){var c=function(){var b=document.getElementById(\"copy-label\");document.getElementById(\"plain-ray-id\");if(navigator.clipboard)navigator.clipboard.writeText(\"70cd75d6ef529681\");else{var a=document.createElement(\"textarea\");a.value=\"70cd75d6ef529681\";a.style.top=\"0\";a.style.left=\"0\";a.style.position=\"fixed\";document.body.appendChild(a);a.focus();a.select();document.execCommand(\"copy\");document.body.removeChild(a)}b.innerText=\"Copied\"};document.addEventListener(\"DOMContentLoaded\",\\nfunction(){var b=document.getElementById(\"plain-ray-id\"),a=document.getElementById(\"click-to-copy-btn\");\"classList\"in b&&(b.classList.add(\"hidden\"),a.classList.remove(\"hidden\"),a.addEventListener(\"click\",c))})}})();\\n      </script>\\n      <script defer src=\"https://api.radar.cloudflare.com/beacon.js\"></script>\\n   </head>\\n\\n   <body>\\n      <div class=\"main-wrapper\" role=\"main\">\\n         <div class=\"header section\">\\n            <h1>\\n               <span class=\"error-description\">Access denied</span>\\n               <span class=\"code-label\">Error code <span>1020</span></span>\\n            </h1>\\n            <div class=\"large-font\">\\n               <p>You do not have access to api.opensea.io.</p><p>The site owner may have set restrictions that prevent you from accessing the site. Contact the site owner for access or try loading the page again.</p>\\n            </div>\\n         </div>\\n      </div>\\n\\n      <div>\\n         <div class=\"clearfix section\">\\n            <div class=\"column know-more\">\\n               <h2 class=\"large-font\">Additional information</h2>\\n               <p>The access policies of a site define which visits are allowed. Your current visit is not allowed according to those policies.</p><p>Only the site owner can change site access policies.</p>\\n            </div>\\n            <div class=\"column know-more\">\\n               <h2 class=\"large-font\">I am the site owner</h2>\\n               <p class=\"ray-id-wrapper\">\\n                  Ray ID:\\n                  <span class=\"plain-ray-id\" id=\"plain-ray-id\">\\n                     70cd75d6ef529681\\n                  </span>\\n                  <button class=\"click-to-copy-btn hidden\" id=\"click-to-copy-btn\" title=\"Click to copy Ray ID\" type=\"button\">\\n                     <span class=\"ray-id\">70cd75d6ef529681</span><span class=\"copy-label\" id=\"copy-label\">Copy</span>\\n                  </button>\\n               </p>\\n               <ol>\\n                  <li>\\n                     Search the\\n                     <a rel=\"noopener noreferrer\" href=\"https://dash.cloudflare.com/?to=/:account/:zone/firewall\" target=\"_blank\">Firewall Events Log</a>\\n                     <img class=\"external-link\" title=\"Opens in new tab\" src=\"/cdn-cgi/images/external.png\" alt=\"External link\">\\n                     for the above Ray ID.\\n                  </li>\\n                  <li>Examine and assess the details of the access policy.</li>\\n               </ol>\\n               <br>\\n               <a rel=\"noopener noreferrer\" href=\"https://support.cloudflare.com/hc/articles/360029779472-Troubleshooting-Cloudflare-1XXX-errors#error1020\" target=\"_blank\">Troubleshooting guide</a>\\n               <img class=\"external-link\" title=\"Opens in new tab\" src=\"/cdn-cgi/images/external.png\" alt=\"External link\">\\n            </div>\\n         </div>\\n\\n         <div class=\"clearfix footer section\" role=\"contentinfo\">\\n            <div class=\"column\">\\n               <div class=\"feedback-hidden py-8 text-center\" id=\"error-feedback\">\\n    <div id=\"error-feedback-survey\" class=\"footer-line-wrapper\">\\n        Was this page helpful?\\n        <button class=\"border border-solid bg-white cf-button cursor-pointer ml-4 px-4 py-2 rounded\" id=\"feedback-button-yes\" type=\"button\">Yes</button>\\n        <button class=\"border border-solid bg-white cf-button cursor-pointer ml-4 px-4 py-2 rounded\" id=\"feedback-button-no\" type=\"button\">No</button>\\n    </div>\\n    <div class=\"feedback-success feedback-hidden\" id=\"error-feedback-success\">\\n        Thank you for your feedback!\\n    </div>\\n</div>\\n\\n            </div>\\n            <div class=\"column footer-line-wrapper text-center\">\\n               Performance &amp; security by <a rel=\"noopener noreferrer\" href=\"https://www.cloudflare.com\" target=\"_blank\">Cloudflare</a>\\n               <img class=\"external-link\" title=\"Opens in new tab\" src=\"/cdn-cgi/images/external.png\" alt=\"External link\">\\n            </div>\\n         </div>\\n      </div>\\n   <script defer src=\"https://static.cloudflareinsights.com/beacon.min.js/v652eace1692a40cfa3763df669d7439c1639079717194\" integrity=\"sha512-Gi7xpJR8tSkrpF7aordPZQlW2DLtzUlZcumS8dMQjwDHEnw9I7ZLyiOj/6tZStRBGtGgN6ceN6cMH8z7etPGlw==\" data-cf-beacon=\\'{\"rayId\":\"70cd75d6ef529681\",\"token\":\"96047490de144c8b91be74ba7605ab69\",\"version\":\"2021.12.0\",\"si\":100}\\' crossorigin=\"anonymous\"></script>\\n</body>\\n</html>\\n'"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response.text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Instantiate API connections \n",
    "nft_price_floor = NFTPriceFloor()\n",
    "non_fungible = NonFungible()\n",
    "opensea = OpenSea()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collections of interest to profile \n",
    "# Names are slightly different for each source \n",
    "\n",
    "nft_price_floor_collections = [\n",
    "    'boredApesYC',\n",
    "    'coolCats',\n",
    "    'cryptoPunks',\n",
    "]\n",
    "\n",
    "non_fungible_collections = [\n",
    "    'cryptopunks',\n",
    "    'doodles',\n",
    "]\n",
    "\n",
    "opensea_collections = [\n",
    "    'doodles-official',\n",
    "]\n",
    "\n",
    "opensea_contract_addresses = [\n",
    "    '0xb47e3cd837ddf8e4c57f05d70ab865de6e193bbb', # punks\n",
    "    '0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D', # BAYC\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "ename": "ConnectionError",
     "evalue": "HTTPSConnectionPool(host='api.nftpricefloor.com', port=443): Max retries exceeded with url: /nfts (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fc988c6c5e0>: Failed to establish a new connection: [Errno 60] Operation timed out'))",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTimeoutError\u001b[0m                              Traceback (most recent call last)",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py:174\u001b[0m, in \u001b[0;36mHTTPConnection._new_conn\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py?line=172'>173</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py?line=173'>174</a>\u001b[0m     conn \u001b[39m=\u001b[39m connection\u001b[39m.\u001b[39;49mcreate_connection(\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py?line=174'>175</a>\u001b[0m         (\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_dns_host, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mport), \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mtimeout, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mextra_kw\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py?line=175'>176</a>\u001b[0m     )\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py?line=177'>178</a>\u001b[0m \u001b[39mexcept\u001b[39;00m SocketTimeout:\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/util/connection.py:95\u001b[0m, in \u001b[0;36mcreate_connection\u001b[0;34m(address, timeout, source_address, socket_options)\u001b[0m\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/util/connection.py?line=93'>94</a>\u001b[0m \u001b[39mif\u001b[39;00m err \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m---> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/util/connection.py?line=94'>95</a>\u001b[0m     \u001b[39mraise\u001b[39;00m err\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/util/connection.py?line=96'>97</a>\u001b[0m \u001b[39mraise\u001b[39;00m socket\u001b[39m.\u001b[39merror(\u001b[39m\"\u001b[39m\u001b[39mgetaddrinfo returns an empty list\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/util/connection.py:85\u001b[0m, in \u001b[0;36mcreate_connection\u001b[0;34m(address, timeout, source_address, socket_options)\u001b[0m\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/util/connection.py?line=83'>84</a>\u001b[0m     sock\u001b[39m.\u001b[39mbind(source_address)\n\u001b[0;32m---> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/util/connection.py?line=84'>85</a>\u001b[0m sock\u001b[39m.\u001b[39;49mconnect(sa)\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/util/connection.py?line=85'>86</a>\u001b[0m \u001b[39mreturn\u001b[39;00m sock\n",
      "\u001b[0;31mTimeoutError\u001b[0m: [Errno 60] Operation timed out",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mNewConnectionError\u001b[0m                        Traceback (most recent call last)",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py:703\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=701'>702</a>\u001b[0m \u001b[39m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=702'>703</a>\u001b[0m httplib_response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_request(\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=703'>704</a>\u001b[0m     conn,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=704'>705</a>\u001b[0m     method,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=705'>706</a>\u001b[0m     url,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=706'>707</a>\u001b[0m     timeout\u001b[39m=\u001b[39;49mtimeout_obj,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=707'>708</a>\u001b[0m     body\u001b[39m=\u001b[39;49mbody,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=708'>709</a>\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=709'>710</a>\u001b[0m     chunked\u001b[39m=\u001b[39;49mchunked,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=710'>711</a>\u001b[0m )\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=712'>713</a>\u001b[0m \u001b[39m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=713'>714</a>\u001b[0m \u001b[39m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=714'>715</a>\u001b[0m \u001b[39m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=715'>716</a>\u001b[0m \u001b[39m# mess.\u001b[39;00m\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py:386\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=384'>385</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=385'>386</a>\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_validate_conn(conn)\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=386'>387</a>\u001b[0m \u001b[39mexcept\u001b[39;00m (SocketTimeout, BaseSSLError) \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=387'>388</a>\u001b[0m     \u001b[39m# Py2 raises this as a BaseSSLError, Py3 raises it as socket timeout.\u001b[39;00m\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py:1040\u001b[0m, in \u001b[0;36mHTTPSConnectionPool._validate_conn\u001b[0;34m(self, conn)\u001b[0m\n\u001b[1;32m   <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=1038'>1039</a>\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mgetattr\u001b[39m(conn, \u001b[39m\"\u001b[39m\u001b[39msock\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m):  \u001b[39m# AppEngine might not have  `.sock`\u001b[39;00m\n\u001b[0;32m-> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=1039'>1040</a>\u001b[0m     conn\u001b[39m.\u001b[39;49mconnect()\n\u001b[1;32m   <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=1041'>1042</a>\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m conn\u001b[39m.\u001b[39mis_verified:\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py:358\u001b[0m, in \u001b[0;36mHTTPSConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py?line=355'>356</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mconnect\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py?line=356'>357</a>\u001b[0m     \u001b[39m# Add certificate verification\u001b[39;00m\n\u001b[0;32m--> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py?line=357'>358</a>\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39msock \u001b[39m=\u001b[39m conn \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_new_conn()\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py?line=358'>359</a>\u001b[0m     hostname \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhost\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py:186\u001b[0m, in \u001b[0;36mHTTPConnection._new_conn\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py?line=184'>185</a>\u001b[0m \u001b[39mexcept\u001b[39;00m SocketError \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m--> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py?line=185'>186</a>\u001b[0m     \u001b[39mraise\u001b[39;00m NewConnectionError(\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py?line=186'>187</a>\u001b[0m         \u001b[39mself\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mFailed to establish a new connection: \u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m\"\u001b[39m \u001b[39m%\u001b[39m e\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py?line=187'>188</a>\u001b[0m     )\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connection.py?line=189'>190</a>\u001b[0m \u001b[39mreturn\u001b[39;00m conn\n",
      "\u001b[0;31mNewConnectionError\u001b[0m: <urllib3.connection.HTTPSConnection object at 0x7fc988c6c5e0>: Failed to establish a new connection: [Errno 60] Operation timed out",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mMaxRetryError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py:440\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=438'>439</a>\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m chunked:\n\u001b[0;32m--> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=439'>440</a>\u001b[0m     resp \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49murlopen(\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=440'>441</a>\u001b[0m         method\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mmethod,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=441'>442</a>\u001b[0m         url\u001b[39m=\u001b[39;49murl,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=442'>443</a>\u001b[0m         body\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mbody,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=443'>444</a>\u001b[0m         headers\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mheaders,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=444'>445</a>\u001b[0m         redirect\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=445'>446</a>\u001b[0m         assert_same_host\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=446'>447</a>\u001b[0m         preload_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=447'>448</a>\u001b[0m         decode_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=448'>449</a>\u001b[0m         retries\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mmax_retries,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=449'>450</a>\u001b[0m         timeout\u001b[39m=\u001b[39;49mtimeout\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=450'>451</a>\u001b[0m     )\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=452'>453</a>\u001b[0m \u001b[39m# Send the request.\u001b[39;00m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=453'>454</a>\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py:785\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=782'>783</a>\u001b[0m     e \u001b[39m=\u001b[39m ProtocolError(\u001b[39m\"\u001b[39m\u001b[39mConnection aborted.\u001b[39m\u001b[39m\"\u001b[39m, e)\n\u001b[0;32m--> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=784'>785</a>\u001b[0m retries \u001b[39m=\u001b[39m retries\u001b[39m.\u001b[39;49mincrement(\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=785'>786</a>\u001b[0m     method, url, error\u001b[39m=\u001b[39;49me, _pool\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m, _stacktrace\u001b[39m=\u001b[39;49msys\u001b[39m.\u001b[39;49mexc_info()[\u001b[39m2\u001b[39;49m]\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=786'>787</a>\u001b[0m )\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/connectionpool.py?line=787'>788</a>\u001b[0m retries\u001b[39m.\u001b[39msleep()\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/util/retry.py:592\u001b[0m, in \u001b[0;36mRetry.increment\u001b[0;34m(self, method, url, response, error, _pool, _stacktrace)\u001b[0m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/util/retry.py?line=590'>591</a>\u001b[0m \u001b[39mif\u001b[39;00m new_retry\u001b[39m.\u001b[39mis_exhausted():\n\u001b[0;32m--> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/util/retry.py?line=591'>592</a>\u001b[0m     \u001b[39mraise\u001b[39;00m MaxRetryError(_pool, url, error \u001b[39mor\u001b[39;00m ResponseError(cause))\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/urllib3/util/retry.py?line=593'>594</a>\u001b[0m log\u001b[39m.\u001b[39mdebug(\u001b[39m\"\u001b[39m\u001b[39mIncremented Retry for (url=\u001b[39m\u001b[39m'\u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m): \u001b[39m\u001b[39m%r\u001b[39;00m\u001b[39m\"\u001b[39m, url, new_retry)\n",
      "\u001b[0;31mMaxRetryError\u001b[0m: HTTPSConnectionPool(host='api.nftpricefloor.com', port=443): Max retries exceeded with url: /nfts (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fc988c6c5e0>: Failed to establish a new connection: [Errno 60] Operation timed out'))",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mConnectionError\u001b[0m                           Traceback (most recent call last)",
      "\u001b[1;32m/Users/alexkan/work/api/messari-python-api-private/examples/notebooks/NftProfiler.ipynb Cell 7'\u001b[0m in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/alexkan/work/api/messari-python-api-private/examples/notebooks/NftProfiler.ipynb#ch0000004?line=0'>1</a>\u001b[0m \u001b[39m# # Getting NFT Price Floor data \u001b[39;00m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/alexkan/work/api/messari-python-api-private/examples/notebooks/NftProfiler.ipynb#ch0000004?line=2'>3</a>\u001b[0m nfts \u001b[39m=\u001b[39m nft_price_floor\u001b[39m.\u001b[39;49mget_nfts()\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/nftpricefloor/nftpricefloor.py:28\u001b[0m, in \u001b[0;36mNFTPriceFloor.get_nfts\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/nftpricefloor/nftpricefloor.py?line=19'>20</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mget_nfts\u001b[39m(\u001b[39mself\u001b[39m) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m pd\u001b[39m.\u001b[39mDataFrame:\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/nftpricefloor/nftpricefloor.py?line=20'>21</a>\u001b[0m     \u001b[39m\"\"\"Retrieve basic info for all nfts\u001b[39;00m\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/nftpricefloor/nftpricefloor.py?line=21'>22</a>\u001b[0m \n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/nftpricefloor/nftpricefloor.py?line=22'>23</a>\u001b[0m \u001b[39m    Returns\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/nftpricefloor/nftpricefloor.py?line=25'>26</a>\u001b[0m \u001b[39m            DataFrame with info on all nfts\u001b[39;00m\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/nftpricefloor/nftpricefloor.py?line=26'>27</a>\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/nftpricefloor/nftpricefloor.py?line=27'>28</a>\u001b[0m     response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mget_response(NFTS_URL)\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/nftpricefloor/nftpricefloor.py?line=28'>29</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m pd\u001b[39m.\u001b[39mDataFrame(response)\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/dataloader.py:52\u001b[0m, in \u001b[0;36mDataLoader.get_response\u001b[0;34m(self, endpoint_url, params, headers)\u001b[0m\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/dataloader.py?line=49'>50</a>\u001b[0m \u001b[39m# Make request\u001b[39;00m\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/dataloader.py?line=50'>51</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m---> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/dataloader.py?line=51'>52</a>\u001b[0m     response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msession\u001b[39m.\u001b[39;49mget(endpoint_url, params\u001b[39m=\u001b[39;49mparams, headers\u001b[39m=\u001b[39;49mheaders)\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/dataloader.py?line=52'>53</a>\u001b[0m \u001b[39mexcept\u001b[39;00m requests\u001b[39m.\u001b[39mexceptions\u001b[39m.\u001b[39mHTTPError \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/dataloader.py?line=53'>54</a>\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mSystemError\u001b[39;00m(e) \u001b[39mfrom\u001b[39;00m \u001b[39me\u001b[39;00m\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py:542\u001b[0m, in \u001b[0;36mSession.get\u001b[0;34m(self, url, **kwargs)\u001b[0m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=533'>534</a>\u001b[0m \u001b[39mr\u001b[39m\u001b[39m\"\"\"Sends a GET request. Returns :class:`Response` object.\u001b[39;00m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=534'>535</a>\u001b[0m \n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=535'>536</a>\u001b[0m \u001b[39m:param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=536'>537</a>\u001b[0m \u001b[39m:param \\*\\*kwargs: Optional arguments that ``request`` takes.\u001b[39;00m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=537'>538</a>\u001b[0m \u001b[39m:rtype: requests.Response\u001b[39;00m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=538'>539</a>\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=540'>541</a>\u001b[0m kwargs\u001b[39m.\u001b[39msetdefault(\u001b[39m'\u001b[39m\u001b[39mallow_redirects\u001b[39m\u001b[39m'\u001b[39m, \u001b[39mTrue\u001b[39;00m)\n\u001b[0;32m--> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=541'>542</a>\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mrequest(\u001b[39m'\u001b[39;49m\u001b[39mGET\u001b[39;49m\u001b[39m'\u001b[39;49m, url, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py:529\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=523'>524</a>\u001b[0m send_kwargs \u001b[39m=\u001b[39m {\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=524'>525</a>\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mtimeout\u001b[39m\u001b[39m'\u001b[39m: timeout,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=525'>526</a>\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mallow_redirects\u001b[39m\u001b[39m'\u001b[39m: allow_redirects,\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=526'>527</a>\u001b[0m }\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=527'>528</a>\u001b[0m send_kwargs\u001b[39m.\u001b[39mupdate(settings)\n\u001b[0;32m--> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=528'>529</a>\u001b[0m resp \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msend(prep, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49msend_kwargs)\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=530'>531</a>\u001b[0m \u001b[39mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py:645\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=641'>642</a>\u001b[0m start \u001b[39m=\u001b[39m preferred_clock()\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=643'>644</a>\u001b[0m \u001b[39m# Send the request\u001b[39;00m\n\u001b[0;32m--> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=644'>645</a>\u001b[0m r \u001b[39m=\u001b[39m adapter\u001b[39m.\u001b[39;49msend(request, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=646'>647</a>\u001b[0m \u001b[39m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/sessions.py?line=647'>648</a>\u001b[0m elapsed \u001b[39m=\u001b[39m preferred_clock() \u001b[39m-\u001b[39m start\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py:519\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=514'>515</a>\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(e\u001b[39m.\u001b[39mreason, _SSLError):\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=515'>516</a>\u001b[0m         \u001b[39m# This branch is for urllib3 v1.22 and later.\u001b[39;00m\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=516'>517</a>\u001b[0m         \u001b[39mraise\u001b[39;00m SSLError(e, request\u001b[39m=\u001b[39mrequest)\n\u001b[0;32m--> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=518'>519</a>\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m(e, request\u001b[39m=\u001b[39mrequest)\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=520'>521</a>\u001b[0m \u001b[39mexcept\u001b[39;00m ClosedPoolError \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/requests/adapters.py?line=521'>522</a>\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m(e, request\u001b[39m=\u001b[39mrequest)\n",
      "\u001b[0;31mConnectionError\u001b[0m: HTTPSConnectionPool(host='api.nftpricefloor.com', port=443): Max retries exceeded with url: /nfts (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fc988c6c5e0>: Failed to establish a new connection: [Errno 60] Operation timed out'))"
     ]
    }
   ],
   "source": [
    "# # Getting NFT Price Floor data \n",
    "\n",
    "nfts = nft_price_floor.get_nfts()\n",
    "# nfts.head()\n",
    "# df_floors = nft_price_floor.get_floor(nft_price_floor_collections)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "ename": "Exception",
     "evalue": "Query Failed - code: 403",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mException\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/Users/alexkan/work/api/messari-python-api-private/examples/notebooks/NftProfiler.ipynb Cell 5'\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/alexkan/work/api/messari-python-api-private/examples/notebooks/NftProfiler.ipynb#ch0000006?line=0'>1</a>\u001b[0m df_collection_history \u001b[39m=\u001b[39m non_fungible\u001b[39m.\u001b[39;49mget_collection_history(collections_in\u001b[39m=\u001b[39;49mnon_fungible_collections, length\u001b[39m=\u001b[39;49m\u001b[39m5\u001b[39;49m)\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/alexkan/work/api/messari-python-api-private/examples/notebooks/NftProfiler.ipynb#ch0000006?line=1'>2</a>\u001b[0m df_collection_stats \u001b[39m=\u001b[39m non_fungible\u001b[39m.\u001b[39mget_collection_stats(collections_in\u001b[39m=\u001b[39mnon_fungible_collections)\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/alexkan/work/api/messari-python-api-private/examples/notebooks/NftProfiler.ipynb#ch0000006?line=2'>3</a>\u001b[0m df_collection_summary \u001b[39m=\u001b[39m non_fungible\u001b[39m.\u001b[39mget_collection_summary(collections_in\u001b[39m=\u001b[39mnon_fungible_collections)\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/nonfungible/nonfungible.py:44\u001b[0m, in \u001b[0;36mNonFungible.get_collection_history\u001b[0;34m(self, collections_in, length)\u001b[0m\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/nonfungible/nonfungible.py?line=41'>42</a>\u001b[0m \u001b[39mfor\u001b[39;00m collection \u001b[39min\u001b[39;00m collections:\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/nonfungible/nonfungible.py?line=42'>43</a>\u001b[0m     endpoint_url \u001b[39m=\u001b[39m COLLECTION_HISTORY_URL\u001b[39m.\u001b[39msubstitute(collection\u001b[39m=\u001b[39mcollection)\n\u001b[0;32m---> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/nonfungible/nonfungible.py?line=43'>44</a>\u001b[0m     response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mget_response(endpoint_url, params\u001b[39m=\u001b[39;49mparams)\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/nonfungible/nonfungible.py?line=44'>45</a>\u001b[0m     tmp_df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame(response[\u001b[39m'\u001b[39m\u001b[39mdata\u001b[39m\u001b[39m'\u001b[39m][\u001b[39m'\u001b[39m\u001b[39msales\u001b[39m\u001b[39m'\u001b[39m][\u001b[39m0\u001b[39m][\u001b[39m'\u001b[39m\u001b[39msales\u001b[39m\u001b[39m'\u001b[39m])\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/nonfungible/nonfungible.py?line=45'>46</a>\u001b[0m     df_list\u001b[39m.\u001b[39mappend(tmp_df)\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/dataloader.py:60\u001b[0m, in \u001b[0;36mDataLoader.get_response\u001b[0;34m(self, endpoint_url, params, headers)\u001b[0m\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/dataloader.py?line=57'>58</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m response\u001b[39m.\u001b[39mjson()\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/dataloader.py?line=58'>59</a>\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m---> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/dataloader.py?line=59'>60</a>\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mException\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mQuery Failed - code: \u001b[39m\u001b[39m{\u001b[39;00mresponse\u001b[39m.\u001b[39mstatus_code\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m)\n",
      "\u001b[0;31mException\u001b[0m: Query Failed - code: 403"
     ]
    }
   ],
   "source": [
    "df_collection_history = non_fungible.get_collection_history(collections_in=non_fungible_collections, length=5)\n",
    "df_collection_stats = non_fungible.get_collection_stats(collections_in=non_fungible_collections)\n",
    "df_collection_summary = non_fungible.get_collection_summary(collections_in=non_fungible_collections)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "Exception",
     "evalue": "Query Failed - code: 403",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mException\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/Users/alexkan/work/api/messari-python-api-private/examples/notebooks/NftProfiler.ipynb Cell 6'\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/alexkan/work/api/messari-python-api-private/examples/notebooks/NftProfiler.ipynb#ch0000007?line=0'>1</a>\u001b[0m df_contracts \u001b[39m=\u001b[39m opensea\u001b[39m.\u001b[39;49mget_contract(contracts_in\u001b[39m=\u001b[39;49mopensea_contract_addresses)\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/opensea/opensea.py:95\u001b[0m, in \u001b[0;36mOpenSea.get_contract\u001b[0;34m(self, contracts_in)\u001b[0m\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/opensea/opensea.py?line=92'>93</a>\u001b[0m \u001b[39mfor\u001b[39;00m contract \u001b[39min\u001b[39;00m contracts:\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/opensea/opensea.py?line=93'>94</a>\u001b[0m     endpoint_url \u001b[39m=\u001b[39m CONTRACT_URL\u001b[39m.\u001b[39msubstitute(contract\u001b[39m=\u001b[39mcontract)\n\u001b[0;32m---> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/opensea/opensea.py?line=94'>95</a>\u001b[0m     response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mget_response(endpoint_url, headers\u001b[39m=\u001b[39;49mheaders)\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/opensea/opensea.py?line=95'>96</a>\u001b[0m     tmp_df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame(response)\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/nfts/opensea/opensea.py?line=97'>98</a>\u001b[0m     df_list\u001b[39m.\u001b[39mappend(tmp_df)\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/dataloader.py:60\u001b[0m, in \u001b[0;36mDataLoader.get_response\u001b[0;34m(self, endpoint_url, params, headers)\u001b[0m\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/dataloader.py?line=57'>58</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m response\u001b[39m.\u001b[39mjson()\n\u001b[1;32m     <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/dataloader.py?line=58'>59</a>\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m---> <a href='file:///Users/alexkan/opt/anaconda3/envs/env/lib/python3.9/site-packages/messari/dataloader.py?line=59'>60</a>\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mException\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mQuery Failed - code: \u001b[39m\u001b[39m{\u001b[39;00mresponse\u001b[39m.\u001b[39mstatus_code\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m)\n",
      "\u001b[0;31mException\u001b[0m: Query Failed - code: 403"
     ]
    }
   ],
   "source": [
    "df_contracts = opensea.get_contract(contracts_in=opensea_contract_addresses)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "7a7ae9f426fd204616fa339a0e58b69e82ffc4e43843a18b6d9b7840d568dd9f"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
